File: objects/examples/github.py
#!/usr/bin/env python
# vim: ft=python ts=4 sw=4 expandtab
#
# date: Wed Jul  6 06:46:15 2016

import os
import sys
import httplib
import json

def getinput(describe):
    sys.stderr.write(describe + ': ')
    line = sys.stdin.readline()
    line = line.rstrip()
    return line

def jsonfile():
    basefile = os.path.basename(__file__)
    return os.path.join(os.path.dirname(__file__),os.path.splitext(basefile)[0] + '.json')

def main(arguments):
    json_path = jsonfile()
    json_out = ''

    if not os.path.exists(json_path):
        resp = None
        while True:
            conn = httplib.HTTPSConnection('api.github.com')
            if not os.path.exists(json_path):
                username = getinput('What is your github username? ')

                conn.request(method='GET', url='https://api.github.com/users/' + username,headers={'User-Agent': 'Codestrap-Github-Metadata'})

                resp = conn.getresponse()

                if resp.status != 200:
                    sys.stderr.write('Invalid response code for https://api.github.com/users/' + username + '. Returned: ' + resp.status + '\n')
                    sys.stderr.write('Please reenter your username\n')
                else:
                    break

        output = json.loads(resp.read())
        json_out = json.dumps(output, indent = 4, separators = (',', ': '))

        with open(json_path, 'w') as fd:
            fd.write(json_out)

    with open(json_path, 'r') as fd:
        lines = fd.readlines()
        for line in lines:
            sys.stdout.write(line)

if __name__ == '__main__':
    sys.exit(main(sys.argv[1:]))

File: objects/examples/github.rb
#!/usr/bin/env ruby
# vim: ft=ruby ts=2 sw=2 expandtab
#
# date: Fri Jul  8 17:52:29 2016

require 'json'
require 'open-uri'

#
# Definitions
#
def getinput(describe)
  STDERR.print("#{describe}: ")
  STDIN.gets.chomp
end

def jsonfile
  File.join(File.dirname($0), "#{File.basename($0, '.*')}.json")
end

#
# Main
#
file = jsonfile

# Check for object/personal.json
correct = false
resp = nil
json_string = ''
unless File.exist?(file)
    begin
      username = getinput 'Enter your GitHub username'
      #username = 'dexterp'
      resp = open("https://api.github.com/users/#{username}",
                  'User-Agent' => 'Codestrap-Github-Metadata')
      if (resp.status[0] == '200') then
        correct = true
      else
        puts "Invalid username https://api.github.com/users/#{username}"
      end
    end until correct
end
resp.each_line do |line|
  json_string << line
end
json_hash = JSON.parse(json_string)

if correct
  File.open(file, 'w') do |file|
    file.puts(JSON.pretty_generate(json_hash))
  end
end

if File.exist?(file)
  File.open(file, 'r') do |file|
    file.each_line do |line|
      print line
    end
  end
end
